scalatest-test/src/test/scala/org/scalatest/enablers/PropCheckerAssertingAsyncSpec.scala

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright 2001-2019 Artima, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.scalatest.enablers
+
+import org.scalatest._
+import org.scalactic.Equality
+import scala.collection.immutable
+import prop.GeneratorDrivenPropertyChecks
+import org.scalactic.anyvals.PosZInt
+import exceptions.TestFailedException
+import OptionValues._
+import scala.concurrent.Future
+
+class PropCheckerAssertingAsyncSpec extends AsyncFunSpec with Matchers with GeneratorDrivenPropertyChecks with LineNumberHelper {
+
+  describe("PropCheckerAsserting") {
+    it("The exception reported by an async forAll's Future should include the exception thrown by the property function evaluation as its cause") {
```

Comment:
Note: we had a long conversation about this test pattern, and the fact that it isn't obvious why it works.  We should state explicitly, in the forAll() documentation, that if the function returns a Future, the next loop will not be tried until after that Future completes.  So the first loop to fail (even in a Future) will be the one returned by forAll().

Reply:
Also worth noting that this means that forAll() over Futures is *never* going to be parallel (regardless of the EC used); don't expect that to help with your test performance.